File: java/src/test/java/io/cucumber/htmlformatter/MessagesToHtmlWriterTest.java
package io.cucumber.htmlformatter;

import io.cucumber.htmlformatter.MessagesToHtmlWriter.Serializer;
import io.cucumber.messages.types.Comment;
import io.cucumber.messages.types.Envelope;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Location;
import io.cucumber.messages.types.TestRunFinished;
import io.cucumber.messages.types.TestRunStarted;
import io.cucumber.messages.types.Timestamp;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.time.Instant;

import static io.cucumber.messages.Convertor.toMessage;
import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MessagesToHtmlWriterTest {

    static final Serializer serializer = Jackson.OBJECT_MAPPER::writeValue;

    private static ByteArrayInputStream createInputStream(String s) {
        return new ByteArrayInputStream(s.getBytes(UTF_8));
    }

    private static String renderAsHtml(Envelope... messages) throws IOException {
        return renderAsHtml(MessagesToHtmlWriter.builder(serializer), messages);
    }

    private static String renderAsHtml(MessagesToHtmlWriter.Builder builder, Envelope... messages) throws IOException {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        try (MessagesToHtmlWriter messagesToHtmlWriter = builder.build(bytes)) {
            for (Envelope message : messages) {
                messagesToHtmlWriter.write(message);
            }
        }

        return bytes.toString(UTF_8);
    }

    @Test
    void it_writes_one_message_to_html() throws IOException {
        Instant timestamp = Instant.ofEpochSecond(10);
        Envelope envelope = Envelope.of(new TestRunStarted(toMessage(timestamp), null));
        String html = renderAsHtml(envelope);
        assertThat(html).contains("window.CUCUMBER_MESSAGES = [{\"testRunStarted\":{\"timestamp\":{\"seconds\":10,\"nanos\":0}}}];");
    }

    @Test
    void it_writes_no_message_to_html() throws IOException {
        String html = renderAsHtml();
        assertThat(html).contains("window.CUCUMBER_MESSAGES = [];");
    }

    @Test
    void it_writes_default_title() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer));
        assertThat(html).contains("<title>Cucumber</title>");
    }

    @Test
    void it_writes_custom_title() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer).title("Custom Title"));
        assertThat(html).contains("<title>Custom Title</title>");
    }

    @Test
    void it_writes_default_icon() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer));
        assertThat(html).contains("<link rel=\"icon\" href=\"data:image/svg+xml;base64,");
    }

    @Test
    void it_writes_custom_icon() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer)
                .icon("https://example.com/logo.svg"));
        assertThat(html).contains("<link rel=\"icon\" href=\"https://example.com/logo.svg\">");
    }

    @Test
    void it_writes_default_css() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer)
                .css(() -> createInputStream("p { color: red; }")));
        assertThat(html).contains("p { color: red; }");
    }

    @Test
    void it_writes_custom_css() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer)
                .customCss(() -> createInputStream("p { color: red; }")));
        assertThat(html).contains("p { color: red; }");
    }

    @Test
    void it_writes_default_script() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer)
                .script(() -> createInputStream("console.log(\"Hello world\");")));
        assertThat(html).contains("console.log(\"Hello world\");");
    }

    @Test
    void it_writes_custom_script() throws IOException {
        String html = renderAsHtml(MessagesToHtmlWriter.builder(serializer)
                .customScript(() -> createInputStream("console.log(\"Hello world\");")));
        assertThat(html).contains("console.log(\"Hello world\");");
    }

    @Test
    void it_throws_when_writing_after_close() throws IOException {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        MessagesToHtmlWriter messagesToHtmlWriter = MessagesToHtmlWriter.builder(serializer).build(bytes);
        messagesToHtmlWriter.close();
        assertThrows(IOException.class, () -> messagesToHtmlWriter.write(
                Envelope.of(new TestRunStarted(new Timestamp(0L, 0), ""))
        ));
    }

    @Test
    void it_can_be_closed_twice() throws IOException {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        MessagesToHtmlWriter messagesToHtmlWriter = MessagesToHtmlWriter.builder(serializer).build(bytes);
        messagesToHtmlWriter.close();
        assertDoesNotThrow(messagesToHtmlWriter::close);
    }

    @Test
    void it_is_idempotent_under_failure_to_close() {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream() {
            @Override
            public void close() throws IOException {
                throw new IOException("Can't close this");
            }
        };
        MessagesToHtmlWriter messagesToHtmlWriter = MessagesToHtmlWriter.builder(serializer).build(bytes);
        assertThrows(IOException.class, messagesToHtmlWriter::close);
        byte[] before = bytes.toByteArray();
        assertDoesNotThrow(messagesToHtmlWriter::close);
        byte[] after = bytes.toByteArray();
        assertArrayEquals(before, after);
    }

    @Test
    void it_writes_two_messages_separated_by_a_comma() throws IOException {
        Envelope testRunStarted = Envelope.of(new TestRunStarted(toMessage(Instant.ofEpochSecond(10)), null));

        Envelope envelope = Envelope.of(new TestRunFinished(null, true, toMessage(Instant.ofEpochSecond(15)), null, null));

        String html = renderAsHtml(testRunStarted, envelope);

        assertThat(html).contains("window.CUCUMBER_MESSAGES = [{\"testRunStarted\":{\"timestamp\":{\"seconds\":10,\"nanos\":0}}},{\"testRunFinished\":{\"success\":true,\"timestamp\":{\"seconds\":15,\"nanos\":0}}}];");
    }

    @Test
    void it_escapes_opening_angle_bracket() throws IOException {
        Envelope envelope = Envelope.of(new GherkinDocument(
                null,
                null,
                singletonList(new Comment(
                        new Location(0, 0),
                        "</script><script>alert('Hello')</script>"
                ))
        ));
        String html = renderAsHtml(envelope);
        assertThat(html).contains(
                "window.CUCUMBER_MESSAGES = [{\"gherkinDocument\":{\"comments\":[{\"location\":{\"line\":0,\"column\":0},\"text\":\"\\x3C/script>\\x3Cscript>alert('Hello')\\x3C/script>\"}]}}];");
    }
}

File: dotnet/Cucumber.HtmlFormatterTest/MessagesToHtmlWriterTest.cs
using static System.Text.Encoding;
using Io.Cucumber.Messages.Types;
using Cucumber.Messages;
using Cucumber.HtmlFormatter;

namespace Cucumber.HtmlFormatterTest;

[TestClass]
public class MessagesToHtmlWriterTest
{
    private static readonly Action<StreamWriter, Envelope> Serializer = (sw, e) =>
    {
        var s = NdjsonSerializer.Serialize(e);
        sw.Write(s);
    };

    private static readonly Func<StreamWriter, Envelope, Task> AsyncSerializer = async (sw, e) =>
    {
        var s = NdjsonSerializer.Serialize(e);
        await sw.WriteAsync(s);
    };

    [TestMethod]
    public async Task LegacySyncApiRendersTheSameAsAsync()
    {
        Envelope sampleEnvelope = Envelope.Create(new TestRunStarted(Converters.ToTimestamp(DateTime.UnixEpoch.AddSeconds(10).ToUniversalTime()), null));
        string expectedHtml = await RenderAsHtmlAsync(sampleEnvelope);

        // ReSharper disable once MethodHasAsyncOverload
        string html = RenderAsHtml(sampleEnvelope);

        Assert.AreEqual(expectedHtml, html, "Expected legacy sync API to render the same HTML as the async API.");
    }

    [TestMethod]
    public async Task ItWritesOneMessageToHtmlAsync()
    {
        DateTime timestamp = DateTime.UnixEpoch.AddSeconds(10).ToUniversalTime();
        Envelope envelope = Envelope.Create(new TestRunStarted(Converters.ToTimestamp(timestamp), null));
        string html = await RenderAsHtmlAsync(envelope);
        StringAssert.Contains(html, "window.CUCUMBER_MESSAGES = [{\"testRunStarted\":{\"timestamp\":{\"seconds\":10,\"nanos\":0}}}];",
            $"Expected html to contain a testRunStarted message, but instead html contained: {html}");
    }

    [TestMethod]
    public async Task ItWritesNoMessageToHtml()
    {
        string html = await RenderAsHtmlAsync();
        StringAssert.Contains(html, "window.CUCUMBER_MESSAGES = [];");
    }

    [TestMethod]
    public async Task ItWritesDefaultTitle() 
    {
        string html = await RenderAsHtmlAsync();
        StringAssert.Contains(html, "<title>Cucumber</title>");
    }

    [TestMethod]
    public async Task ItWritesCustomTitle()
    {
        string html = await RenderAsHtmlAsync(new HtmlReportSettings { Title = "Custom Title" });
        StringAssert.Contains(html, "<title>Custom Title</title>");
    }

    [TestMethod]
    public async Task ItWritesDefaultIcon()
    {
        string html = await RenderAsHtmlAsync();
        StringAssert.Contains(html, "<link rel=\"icon\" href=\"data:image/svg+xml;base64,");
    }

    [TestMethod]
    public async Task ItWritesCustomIcon()
    {
        string html = await RenderAsHtmlAsync(new HtmlReportSettings { Icon = "<link rel=\"icon\" href=\"https://example.com/logo.svg\">" });
        StringAssert.Contains(html, "<link rel=\"icon\" href=\"https://example.com/logo.svg\">");
    }

    [TestMethod]
    public async Task ItWritesCustomCss()
    {
        string html = await RenderAsHtmlAsync(new HtmlReportSettings { CustomCss = "p { color: red; }" });
        StringAssert.Contains(html, "p { color: red; }");
    }

    [TestMethod]
    public async Task ItWritesCustomScript()
    {
        string html = await RenderAsHtmlAsync(new HtmlReportSettings() { CustomScript = "console.log(\"Hello world\");" });
        StringAssert.Contains(html, "console.log(\"Hello world\");");
    }

    [TestMethod]
    public async Task ItWritesNoMessageToHtmlAsync()
    {
        string html = await RenderAsHtmlAsync();
        StringAssert.Contains(html, "window.CUCUMBER_MESSAGES = [];");
    }

    [TestMethod]
    public void ItThrowsWhenWritingAfterClose()
    {
        MemoryStream bytes = new MemoryStream();
#pragma warning disable CS0618 // Type or member is obsolete
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, Serializer);
#pragma warning restore CS0618 // Type or member is obsolete
        messagesToHtmlWriter.Dispose();
        Assert.ThrowsExactly<IOException>(() => messagesToHtmlWriter.Write(Envelope.Create(new TestRunStarted(Converters.ToTimestamp(DateTime.UnixEpoch.AddSeconds(10).ToUniversalTime()), null))));
    }

    [TestMethod]
    public async Task ItThrowsWhenWritingAfterCloseAsync()
    {
        MemoryStream bytes = new MemoryStream();
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, AsyncSerializer);
        await messagesToHtmlWriter.DisposeAsync();
        await Assert.ThrowsExactlyAsync<IOException>(async () =>
            await messagesToHtmlWriter.WriteAsync(
                Envelope.Create(new TestRunStarted(
                    Converters.ToTimestamp(DateTime.UnixEpoch.AddSeconds(10).ToUniversalTime()), null))));
    }

    [TestMethod]
    public void ItCanBeClosedTwice()
    {
        MemoryStream bytes = new MemoryStream();
#pragma warning disable CS0618 // Type or member is obsolete
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, Serializer);
#pragma warning restore CS0618 // Type or member is obsolete
        messagesToHtmlWriter.Dispose();
        try { messagesToHtmlWriter.Dispose(); }
        catch (System.Exception e)
        {
            throw new System.Exception($"Expected no exception, but got {e.Message}");
        }
    }

    [TestMethod]
    public async Task ItCanBeClosedTwiceAsync()
    {
        MemoryStream bytes = new MemoryStream();
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, AsyncSerializer);
        await messagesToHtmlWriter.DisposeAsync();
        try { await messagesToHtmlWriter.DisposeAsync(); }
        catch (System.Exception e)
        {
            throw new System.Exception($"Expected no exception, but got {e.Message}");
        }
    }

    internal class StreamThatFailsToClose : MemoryStream
    {
        public override void Close()
        {
            throw new System.Exception("Can't close this");
        }
        public override Task FlushAsync(CancellationToken cancellationToken)
        {
            throw new System.Exception("Can't flush this");
        }
    }

    [TestMethod]
    public void ItIsIdempotentUnderFailureToClose()
    {
        var bytes = new StreamThatFailsToClose();
#pragma warning disable CS0618 // Type or member is obsolete
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, Serializer);
#pragma warning restore CS0618 // Type or member is obsolete
        // ReSharper disable once AccessToDisposedClosure
        Assert.ThrowsExactly<System.Exception>(() => messagesToHtmlWriter.Dispose());
        byte[] before = bytes.ToArray();
        try { messagesToHtmlWriter.Dispose(); }
        catch (System.Exception e)
        {
            throw new System.Exception($"Expected no exception, but got {e.Message}");
        }
        byte[] after = bytes.ToArray();
        CollectionAssert.AreEqual(before, after);
    }

    [TestMethod]
    public async Task ItIsIdempotentUnderFailureToCloseAsync()
    {
        var bytes = new StreamThatFailsToClose();
        MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, AsyncSerializer);
        await Assert.ThrowsExactlyAsync<System.Exception>(async () => await messagesToHtmlWriter.DisposeAsync());
        byte[] before = bytes.ToArray();
        try { await messagesToHtmlWriter.DisposeAsync(); }
        catch (System.Exception e)
        {
            throw new System.Exception($"Expected no exception, but got {e.Message}");
        }
        byte[] after = bytes.ToArray();
        CollectionAssert.AreEqual(before, after);
    }

    [TestMethod]
    public async Task ItWritesTwoMessagesSeparatedByACommaAsync()
    {
        Envelope testRunStarted = Envelope.Create(new TestRunStarted(Converters.ToTimestamp(DateTime.UnixEpoch.AddSeconds(10).ToUniversalTime()), null));
        Envelope envelope = Envelope.Create(new TestRunFinished(null, true, Converters.ToTimestamp(DateTime.UnixEpoch.AddSeconds(15).ToUniversalTime()), null, null));
        string html = await RenderAsHtmlAsync(testRunStarted, envelope);
        StringAssert.Contains(html, "window.CUCUMBER_MESSAGES = [{\"testRunStarted\":{\"timestamp\":{\"seconds\":10,\"nanos\":0}}},{\"testRunFinished\":{\"success\":true,\"timestamp\":{\"seconds\":15,\"nanos\":0}}}];");
    }

    [TestMethod]
    public async Task ItEscapesOpeningAngleBracketAsync()
    {
        Envelope envelope = Envelope.Create(new GherkinDocument(
            null,
            null,
            [new(new Location(0L, 0L), "</script><script>alert('Hello')</script>")]
        ));
        string html = await RenderAsHtmlAsync(envelope);
        StringAssert.Contains(html, "window.CUCUMBER_MESSAGES = [{\"gherkinDocument\":{\"comments\":[{\"location\":{\"line\":0,\"column\":0},\"text\":\"\\x3C/script>\\x3Cscript>alert('Hello')\\x3C/script>\"}]}}];"}]}}];",
            "Expected \"window.CUCUMBER_MESSAGES = [{{\\\"gherkinDocument\\\":{{\\\"comments\\\":[{{\\\"location\\\":{{\\\"line\\\":0,\\\"column\\\":0}},\\\"text\\\":\\\"\\\\x3C/script>\\\\x3Cscript>alert('Hello')\\\\x3C/script>\\\"}}]}}];" +
                        $"\nbut instead had: \n{html.Substring(html.IndexOf("window.CUCUMBER"))}");
    }

    private static string RenderAsHtml(params Envelope[] messages)
    {
        MemoryStream bytes = new MemoryStream();
#pragma warning disable CS0618 // Type or member is obsolete
        using (MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, Serializer))
#pragma warning restore CS0618 // Type or member is obsolete
        {
            foreach (Envelope message in messages)
            {
                messagesToHtmlWriter.Write(message);
            }
        }
        return UTF8.GetString(bytes.ToArray());
    }

    private static async Task<string> RenderAsHtmlAsync(params Envelope[] messages)
    {
        return await RenderAsHtmlAsync(null, messages);
    }

    private static async Task<string> RenderAsHtmlAsync(HtmlReportSettings? settings, params Envelope[] messages)
    {
        MemoryStream bytes = new MemoryStream();
        await using (MessagesToHtmlWriter messagesToHtmlWriter = new MessagesToHtmlWriter(bytes, AsyncSerializer, settings))
        {
            foreach (Envelope message in messages)
            {
                await messagesToHtmlWriter.WriteAsync(message);
            }
            await messagesToHtmlWriter.DisposeAsync();
        }
        return UTF8.GetString(bytes.ToArray());
    }
}